vllm/worker/multi_step_worker.py

```diff
@@ -0,0 +1,158 @@
+        self.multi_step_states: List[
+            Optional[MultiStepState]] = [None] * pipeline_parallel_size
+        self.temp_output = None
+
+    def _get_driver_input_and_broadcast(
+        self, execute_model_req: ExecuteModelRequest
+    ) -> Tuple[BroadcastableModelInput, WorkerInput]:
+        """
+        Get the driver input and broadcast it to other workers.
+        """
+        assert self.is_driver_worker
+        virtual_engine = execute_model_req.virtual_engine
+        is_first_multi_step = execute_model_req.is_first_multi_step
+        if is_first_multi_step:
+            worker_input: WorkerInput = self.prepare_worker_input(
+                execute_model_req=execute_model_req)
+            model_input: MutableModelInputForGPUWithMultiStepMetadata = (
+                self.model_runner.prepare_model_input(
+                    execute_model_req.seq_group_metadata_list,
+                    execute_model_req.virtual_engine,
+                    execute_model_req.finished_requests_ids))
+        else:
+            multi_step_state = self.multi_step_states[virtual_engine]
+            worker_input = multi_step_state.worker_input
+            model_input = multi_step_state.model_input
+            frozen_model_input = model_input.frozen_model_input
+            assert frozen_model_input is not None
+            assert frozen_model_input.attn_metadata is not None
+            # clear the cached decode metadata so that it can be recomputed
+            frozen_model_input.attn_metadata._cached_decode_metadata = None
+
+        model_input.is_first_multi_step = is_first_multi_step
+        model_input.is_last_step = execute_model_req.is_last_step
+
+        # we broadcast the last sampled token ids to all TP workers so they can
+        # update their model input metadata inplace.
+        if not is_first_multi_step:
+            if get_pp_group().is_last_rank:
+                assert model_input.outputs[
+                    -1].sampler_output.sampled_token_ids is None
+                assert model_input.outputs[-1].sampled_token_ids is not None
+                model_input.last_sampled_token_ids = model_input.outputs[
+                    -1].sampled_token_ids
+                # free sampled token ids from the previous step if it has been
+                # pythonized. Cannot free the last sampled token ids because
+                # we need it for GPU advance_step.
+                for output in model_input.outputs[:-1]:
+                    if output.pythonized:
+                        output.sampled_token_ids = None
+            else:
+                # otherwise we need to get the cached sampled token ids from the
+                # execute_model_req
+                assert execute_model_req.last_sampled_token_ids is not None
+                model_input.last_sampled_token_ids = (
+                    execute_model_req.last_sampled_token_ids.cuda())
+                model_input.add_sampler_output(
```

Comment:
What is the reason for having add_sampler_output() call here and not in the "if case" (last rank, after line 64)?

Reply:
For every `MultiStepModelRunner` (TP/PP), `advance_step()` updates metadata using the sampled_token_ids from  `model_inputs.outputs[-1]`. `add_sampler_output()` is used to make sure the non-last PP stages has consistent list of SamplerOutputs as the last rank. Since only the last rank actually samples and appends to `model_inputs.outputs` within `MultiStepModelRunner` holding sampled_token_ids before beginning the current step. In short, the last rank already has the previous step appended.